internal/provider/migrate.go

```diff
@@ -0,0 +1,330 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/mitchellh/cli"
+	"github.com/yuin/goldmark/ast"
+	"github.com/yuin/goldmark/text"
+)
+
+var (
+	exampleImportCodeTemplate = "{{codefile \"shell\" \"%s\"}}"
+	exampleTFCodeTemplate     = "{{tffile \"%s\"}}"
+)
+
+type migrator struct {
+	// providerDir is the absolute path to the root provider directory
+	providerDir string
+
+	oldWebsiteDir string
+	templatesDir  string
+	examplesDir   string
+
+	ui cli.Ui
+}
+
+func (m *migrator) infof(format string, a ...interface{}) {
+	m.ui.Info(fmt.Sprintf(format, a...))
+}
+
+func (m *migrator) warnf(format string, a ...interface{}) {
+	m.ui.Warn(fmt.Sprintf(format, a...))
+}
+
+func Migrate(ui cli.Ui, providerDir string, oldWebsiteDir string, templatesDir string, examplesDir string) error {
+	// Ensure provider directory is resolved absolute path
+	if providerDir == "" {
+		wd, err := os.Getwd()
+
+		if err != nil {
+			return fmt.Errorf("error getting working directory: %w", err)
+		}
+
+		providerDir = wd
+	} else {
+		absProviderDir, err := filepath.Abs(providerDir)
+
+		if err != nil {
+			return fmt.Errorf("error getting absolute path with provider directory %q: %w", providerDir, err)
+		}
+
+		providerDir = absProviderDir
+	}
+
+	// Verify provider directory
+	providerDirFileInfo, err := os.Stat(providerDir)
+
+	if err != nil {
+		return fmt.Errorf("error getting information for provider directory %q: %w", providerDir, err)
+	}
+
+	if !providerDirFileInfo.IsDir() {
+		return fmt.Errorf("expected %q to be a directory", providerDir)
+	}
+
+	m := &migrator{
+		providerDir:   providerDir,
+		oldWebsiteDir: oldWebsiteDir,
+		templatesDir:  templatesDir,
+		examplesDir:   examplesDir,
+
+		ui: ui,
+	}
+
+	return m.Migrate()
+}
+
+func (m *migrator) Migrate() error {
+	m.infof("migrating website from %q to %q", m.OldProviderWebsiteDir(), m.ProviderTemplatesDir())
+
+	err := filepath.Walk(m.OldProviderWebsiteDir(), func(path string, info os.FileInfo, _ error) error {
+		if info.IsDir() {
+			// skip directories
+			return nil
+		}
+
+		rel, err := filepath.Rel(filepath.Join(m.OldProviderWebsiteDir()), path)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve the relative path of basepath %q and targetpath %q: %w",
+				filepath.Join(m.OldProviderWebsiteDir()), path, err)
+		}
+
+		relDir, relFile := filepath.Split(rel)
+		relDir = filepath.ToSlash(relDir)
+
+		data, err := os.ReadFile(path)
+		if err != nil {
+			return fmt.Errorf("unable to read file %q: %w", rel, err)
+		}
+
+		switch relDir {
+		case "docs/d/": //data-sources
+			datasourceName := strings.TrimSuffix(relFile, ".html.markdown")
```

Comment:
The public registry confusingly also supports basically any file extension for legacy documentation (e.g. `.html.md`, `.markdown`, `.md`).

Here's some examples:

- https://github.com/hashicorp/terraform-provider-vault/tree/main/website/docs/r (ad_secret_backend.html.md - [rendered](https://registry.terraform.io/providers/hashicorp/vault/latest/docs/resources/ad_secret_backend), database_secret_backend_role.md - [rendered](https://registry.terraform.io/providers/hashicorp/vault/latest/docs/resources/database_secret_backend_role))
- https://github.com/newrelic/terraform-provider-newrelic/tree/main/website/docs/r (synthetics_multilocation_alert_condition.markdown - [rendered](https://registry.terraform.io/providers/newrelic/newrelic/latest/docs/resources/synthetics_multilocation_alert_condition))

We'll need to ensure we're getting these for all these old providers. 👍